File: android/src/main/kotlin/com/paylisher/flutter/PaylisherFlutterPlugin.kt
package com.paylisher.flutter

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import com.paylisher.Paylisher
import com.paylisher.PersonProfiles
import com.paylisher.android.PaylisherAndroid
import com.paylisher.android.PaylisherAndroidConfig
import com.paylisher.android.internal.getApplicationInfo
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.Date

class PaylisherFlutterPlugin :
    FlutterPlugin,
    MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var applicationContext: Context

    private val snapshotSender = SnapshotSender()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "paylisher_flutter")
        applicationContext = binding.applicationContext
        Log.d("Paylisher", "[FlutterPlugin] onAttachedToEngine() çağırıldı")

        initPlugin()

        channel.setMethodCallHandler(this)
    }

    private fun initPlugin() {
        try {
            val ai = getApplicationInfo(applicationContext)
            val bundle = ai.metaData ?: Bundle()
            val autoInit = bundle.getBoolean("com.paylisher.paylisher.AUTO_INIT", true)

            if (!autoInit) {
                Log.i("Paylisher", "[initPlugin] AUTO_INIT disabled")
                return
            }

            val apiKey = bundle.getString("com.paylisher.paylisher.API_KEY")
            if (apiKey.isNullOrEmpty()) {
                Log.e("Paylisher", "[initPlugin] API_KEY missing")
                return
            }

            val host = bundle.getString("com.paylisher.paylisher.PAYLISHER_HOST")
                ?: "https://datastudio.paylisher.com"

            val captureLifecycle = bundle.getBoolean(
                "com.paylisher.paylisher.TRACK_APPLICATION_LIFECYCLE_EVENTS",
                false
            )
            val debug = bundle.getBoolean("com.paylisher.paylisher.DEBUG", false)

            val cfg = mutableMapOf<String, Any>()
            cfg["apiKey"] = apiKey
            cfg["host"] = host
            cfg["captureApplicationLifecycleEvents"] = captureLifecycle
            cfg["debug"] = debug

            Log.d("Paylisher", "[initPlugin] Auto setup başlatılıyor…")

            setupPaylisher(cfg)

        } catch (e: Throwable) {
            Log.e("Paylisher", "[initPlugin] Hata: $e")
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        Log.d("Paylisher", "[FlutterPlugin] Method call: ${call.method}")

        when (call.method) {
            "setup" -> setup(call, result)
            "identify" -> identify(call, result)
            "capture" -> capture(call, result)
            "screen" -> screen(call, result)
            "alias" -> alias(call, result)
            "distinctId" -> distinctId(result)
            "reset" -> reset(result)
            "disable" -> disable(result)
            "enable" -> enable(result)
            "isOptOut" -> isOptOut(result)
            "isFeatureEnabled" -> isFeatureEnabled(call, result)
            "reloadFeatureFlags" -> reloadFeatureFlags(result)
            "group" -> group(call, result)
            "getFeatureFlag" -> getFeatureFlag(call, result)
            "getFeatureFlagPayload" -> getFeatureFlagPayload(call, result)
            "register" -> register(call, result)
            "unregister" -> unregister(call, result)
            "debug" -> debug(call, result)
            "flush" -> flush(result)
            "captureException" -> captureException(call, result)
            "close" -> close(result)
            "sendMetaEvent" -> handleMetaEvent(call, result)
            "sendFullSnapshot" -> handleSendFullSnapshot(call, result)
            "isSessionReplayActive" -> result.success(Paylisher.isSessionReplayActive())
            "getSessionId" -> getSessionId(result)
            "openUrl" -> openUrl(call, result)

            "surveyAction" -> {
                Log.w("Paylisher", "[Survey] Android SDK survey desteklemiyor → yok sayılıyor")
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun setup(call: MethodCall, result: Result) {
        try {
            val args = call.arguments as? Map<String, Any> ?: emptyMap()
            if (args.isEmpty()) {
                result.error("PaylisherFlutterException", "Arguments null", null)
                return
            }

            Log.d("Paylisher", "[setup] Flutter'dan config alındı → $args")

            setupPaylisher(args)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun setupPaylisher(cfg: Map<String, Any>) {
        val apiKey = cfg["apiKey"] as? String
        if (apiKey.isNullOrEmpty()) {
            Log.e("Paylisher", "[setupPaylisher] API Key missing!")
            return
        }

        val host = cfg["host"] as? String ?: "https://datastudio.paylisher.com"
        val config = PaylisherAndroidConfig(apiKey, host).apply {

            cfg.getIfNotNull<Boolean>("captureApplicationLifecycleEvents") {
                captureApplicationLifecycleEvents = it
            }
            cfg.getIfNotNull<Boolean>("debug") {
                debug = it
            }

            // Session Replay
            cfg.getIfNotNull<Boolean>("sessionReplay") {
                sessionReplay = it
            }
            this.sessionReplayConfig.captureLogcat = false

            // Person profiles
            cfg.getIfNotNull<String>("personProfiles") {
                personProfiles = when (it) {
                    "never" -> PersonProfiles.NEVER
                    "always" -> PersonProfiles.ALWAYS
                    "identifiedOnly" -> PersonProfiles.IDENTIFIED_ONLY
                    else -> PersonProfiles.IDENTIFIED_ONLY
                }
            }

            // Surveys — Android desteklemez
            cfg.getIfNotNull<Boolean>("surveys") { enabled ->
                Log.w("Paylisher", "[setupPaylisher] Surveys = $enabled ancak Android SDK desteklemiyor.")
            }

            sdkName = "paylisher-flutter"
            sdkVersion = cfg["sdkVersion"] as? String ?: "1.0.0"
        }

        Log.d("Paylisher", "[setupPaylisher] PaylisherAndroid.setup() çağırılıyor")
        PaylisherAndroid.setup(applicationContext, config)
    }

    private fun handleMetaEvent(call: MethodCall, result: Result) {
        try {
            val width = call.argument<Int>("width") ?: 0
            val height = call.argument<Int>("height") ?: 0
            val screen = call.argument<String>("screen") ?: ""

            snapshotSender.sendMetaEvent(width, height, screen)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun handleSendFullSnapshot(call: MethodCall, result: Result) {
        try {
            val bytes = call.argument<ByteArray>("imageBytes")
            val id = call.argument<Int>("id") ?: 1
            val x = call.argument<Int>("x") ?: 0
            val y = call.argument<Int>("y") ?: 0

            if (bytes != null) {
                snapshotSender.sendFullSnapshot(bytes, id, x, y)
            }

            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun identify(call: MethodCall, result: Result) {
        try {
            val userId = call.argument<String>("userId")!!
            val props = call.argument<Map<String, Any>>("userProperties")
            val once = call.argument<Map<String, Any>>("userPropertiesSetOnce")

            Paylisher.identify(userId, props, once)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun capture(call: MethodCall, result: Result) {
        try {
            val event = call.argument<String>("eventName")!!
            val props = call.argument<Map<String, Any>>("properties")
            Paylisher.capture(event, properties = props)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun screen(call: MethodCall, result: Result) {
        try {
            val name = call.argument<String>("screenName")!!
            val props = call.argument<Map<String, Any>>("properties")
            Paylisher.screen(name, props)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun alias(call: MethodCall, result: Result) {
        try {
            val a = call.argument<String>("alias")!!
            Paylisher.alias(a)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun distinctId(result: Result) {
        try {
            result.success(Paylisher.distinctId())
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun reset(result: Result) {
        try {
            Paylisher.reset()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun enable(result: Result) {
        try {
            Paylisher.optIn()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun disable(result: Result) {
        try {
            Paylisher.optOut()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun isOptOut(result: Result) {
        try {
            result.success(Paylisher.isOptOut())
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun isFeatureEnabled(call: MethodCall, result: Result) {
        try {
            val key = call.argument<String>("key")!!
            result.success(Paylisher.isFeatureEnabled(key))
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun reloadFeatureFlags(result: Result) {
        try {
            Paylisher.reloadFeatureFlags()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun group(call: MethodCall, result: Result) {
        try {
            val type = call.argument<String>("groupType")!!
            val key = call.argument<String>("groupKey")!!
            val props = call.argument<Map<String, Any>>("groupProperties")

            Paylisher.group(type, key, props)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun getFeatureFlag(call: MethodCall, result: Result) {
        try {
            val key = call.argument<String>("key")!!
            result.success(Paylisher.getFeatureFlag(key))
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun getFeatureFlagPayload(call: MethodCall, result: Result) {
        try {
            val key = call.argument<String>("key")!!
            result.success(Paylisher.getFeatureFlagPayload(key))
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun register(call: MethodCall, result: Result) {
        try {
            val key = call.argument<String>("key")!!
            val value = call.argument<Any>("value")!!
            Paylisher.register(key, value)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun unregister(call: MethodCall, result: Result) {
        try {
            val key = call.argument<String>("key")!!
            Paylisher.unregister(key)
            result.success(null)

        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun debug(call: MethodCall, result: Result) {
        try {
            val enabled = call.argument<Boolean>("debug") ?: false
            Paylisher.debug(enabled)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun flush(result: Result) {
        try {
            Paylisher.flush()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun captureException(call: MethodCall, result: Result) {
        try {
            val args = call.arguments as? Map<String, Any> ?: emptyMap()
            val props = args["properties"] as? Map<String, Any>

            Paylisher.capture("\$exception", properties = props)
            result.success(null)

        } catch (e: Throwable) {
            result.error("CAPTURE_EXCEPTION_ERROR", e.message, null)
        }
    }

    private fun close(result: Result) {
        try {
            Paylisher.close()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun getSessionId(result: Result) {
        try {
            val id = Paylisher.getSessionId()
            result.success(id?.toString())
        } catch (e: Throwable) {
            result.error("PaylisherFlutterException", e.message, null)
        }
    }

    private fun openUrl(call: MethodCall, result: Result) {
        try {
            val raw = call.arguments as? String ?: ""
            if (raw.isBlank()) {
                result.error("InvalidArguments", "URL is empty", null)
                return
            }

            val uri = Uri.parse(
                if (raw.startsWith("http")) raw else "https://$raw"
            )

            val intent = Intent(Intent.ACTION_VIEW, uri).apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                addCategory(Intent.CATEGORY_BROWSABLE)
            }

            applicationContext.startActivity(intent)
            result.success(null)

        } catch (e: ActivityNotFoundException) {
            result.error("ActivityNotFound", e.message, null)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> Map<String, Any>.getIfNotNull(
        key: String,
        block: (T) -> Unit,
    ) {
        (this[key] as? T)?.let(block)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        Log.d("Paylisher", "[FlutterPlugin] Detached")
        channel.setMethodCallHandler(null)
    }
}

File: ios/Classes/PaylisherFlutterPlugin.swift
import Paylisher
#if os(iOS)
    import Flutter
    import UIKit
#elseif os(macOS)
    import AppKit
    import FlutterMacOS
#endif

public class PaylisherFlutterPlugin: NSObject, FlutterPlugin {
    private static var instance: PaylisherFlutterPlugin?

    public static func getInstance() -> PaylisherFlutterPlugin? {
        instance
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        #if os(iOS)
            let channel = FlutterMethodChannel(name: "paylisher_flutter", binaryMessenger: registrar.messenger())
            let eventChannel = FlutterEventChannel(name: "paylisher_flutter_events", binaryMessenger: registrar.messenger())
            eventChannel.setStreamHandler(PaylisherFlutterNotificationManager.shared)
        #elseif os(macOS)
            let channel = FlutterMethodChannel(name: "paylisher_flutter", binaryMessenger: registrar.messenger)
        #endif
        let instance = PaylisherFlutterPlugin()
        instance.channel = channel
        PaylisherFlutterPlugin.instance = instance
        initPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }

    private let dispatchQueue = DispatchQueue(label: "com.paylisher.PaylisherFlutterPlugin",
                                              target: .global(qos: .utility))

    private var channel: FlutterMethodChannel?

    public static func initPlugin() {
        let autoInit = Bundle.main.object(forInfoDictionaryKey: "com.paylisher.paylisher.AUTO_INIT") as? Bool ?? true
        if !autoInit {
            print("[Paylisher] com.paylisher.paylisher.AUTO_INIT is disabled!")
            return
        }

        let apiKey = Bundle.main.object(forInfoDictionaryKey: "com.paylisher.paylisher.API_KEY") as? String ?? ""

        let host = Bundle.main.object(forInfoDictionaryKey: "com.paylisher.paylisher.PAYLISHER_HOST") as? String ?? PaylisherConfig.defaultHost
        let captureApplicationLifecycleEvents = Bundle.main.object(forInfoDictionaryKey: "com.paylisher.paylisher.CAPTURE_APPLICATION_LIFECYCLE_EVENTS") as? Bool ?? false
        let debug = Bundle.main.object(forInfoDictionaryKey: "com.paylisher.paylisher.DEBUG") as? Bool ?? false

        setupPaylisher([
            "apiKey": apiKey,
            "host": host,
            "captureApplicationLifecycleEvents": captureApplicationLifecycleEvents,
            "debug": debug,
        ])
    }

    private static func setupPaylisher(_ paylisherConfig: [String: Any]) {
        guard let instance = PaylisherFlutterPlugin.instance else {
            print("[Paylisher] Plugin instance not found!")
            return
        }
        let apiKey = paylisherConfig["apiKey"] as? String ?? ""
        if apiKey.isEmpty {
            print("[Paylisher] apiKey is missing!")
            return
        }

        let host = paylisherConfig["host"] as? String ?? PaylisherConfig.defaultHost

        let config = PaylisherConfig(
            apiKey: apiKey,
            host: host
        )
        config.captureScreenViews = false

        if let captureApplicationLifecycleEvents = paylisherConfig["captureApplicationLifecycleEvents"] as? Bool {
            config.captureApplicationLifecycleEvents = captureApplicationLifecycleEvents
        }
        if let debug = paylisherConfig["debug"] as? Bool {
            config.debug = debug
        }
        if let flushAt = paylisherConfig["flushAt"] as? Int {
            config.flushAt = flushAt
        }
        if let maxQueueSize = paylisherConfig["maxQueueSize"] as? Int {
            config.maxQueueSize = maxQueueSize
        }
        if let maxBatchSize = paylisherConfig["maxBatchSize"] as? Int {
            config.maxBatchSize = maxBatchSize
        }
        if let flushInterval = paylisherConfig["flushInterval"] as? Int {
            config.flushIntervalSeconds = Double(flushInterval)
        }
        if let sendFeatureFlagEvents = paylisherConfig["sendFeatureFlagEvents"] as? Bool {
            config.sendFeatureFlagEvent = sendFeatureFlagEvents
        }
        if let preloadFeatureFlags = paylisherConfig["preloadFeatureFlags"] as? Bool {
            config.preloadFeatureFlags = preloadFeatureFlags
        }
        if let optOut = paylisherConfig["optOut"] as? Bool {
            config.optOut = optOut
        }

        if let personProfiles = paylisherConfig["personProfiles"] as? String {
            switch personProfiles {
            case "never":
                config.personProfiles = .never
            case "always":
                config.personProfiles = .always
            case "identifiedOnly":
                config.personProfiles = .identifiedOnly
            default:
                break
            }
        }
        if let dataMode = paylisherConfig["dataMode"] as? String {
            switch dataMode {
            case "wifi":
                config.dataMode = .wifi
            case "cellular":
                config.dataMode = .cellular
            case "any":
                config.dataMode = .any
            default:
                break
            }
        }
        #if os(iOS)
            // configure session replay
            if let sessionReplay = paylisherConfig["sessionReplay"] as? Bool {
                config.sessionReplay = sessionReplay
            }
            // disabled since Dart has native libs such as http/dio and dont use the ios URLSession
            config.sessionReplayConfig.captureNetworkTelemetry = false
        #endif

        // Update SDK name and version
        paylisherSdkName = "paylisher-flutter"
        paylisherVersion = paylisherFlutterVersion

        PaylisherSDK.shared.setup(config)
        
        #if os(iOS)
        if let instance = PaylisherFlutterPlugin.instance {
            PaylisherSDK.shared.setDeepLinkHandler(instance)
        }
        #endif
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setup":
            setup(call, result: result)
        case "getFeatureFlag":
            getFeatureFlag(call, result: result)
        case "isFeatureEnabled":
            isFeatureEnabled(call, result: result)
        case "getFeatureFlagPayload":
            getFeatureFlagPayload(call, result: result)
        case "identify":
            identify(call, result: result)
        case "capture":
            capture(call, result: result)
        case "screen":
            screen(call, result: result)
        case "alias":
            alias(call, result: result)
        case "distinctId":
            distinctId(result)
        case "reset":
            reset(result)
        case "enable":
            enable(result)
        case "disable":
            disable(result)
        case "isOptOut":
            isOptOut(result)
        case "debug":
            debug(call, result: result)
        case "reloadFeatureFlags":
            reloadFeatureFlags(result)
        case "group":
            group(call, result: result)
        case "register":
            register(call, result: result)
        case "unregister":
            unregister(call, result: result)
        case "flush":
            flush(result)
        case "captureException":
            captureException(call, result: result)
        case "close":
            close(result)
        case "sendMetaEvent":
            sendMetaEvent(call, result: result)
        case "sendFullSnapshot":
            sendFullSnapshot(call, result: result)
        case "isSessionReplayActive":
            isSessionReplayActive(result: result)
        case "getSessionId":
            getSessionId(result: result)
        case "openUrl":
            openUrl(call, result: result)
        case "requestNotificationPermission":
            requestNotificationPermission(result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension PaylisherFlutterPlugin {
    private func sendMetaEvent(_ call: FlutterMethodCall,
                               result: @escaping FlutterResult)
    {
        #if os(iOS)
            let date = Date()
            let timestamp = dateToMillis(date)
            if let args = call.arguments as? [String: Any] {
                let width = args["width"] as? Int ?? 0
                let height = args["height"] as? Int ?? 0
                let screen = args["screen"] as? String ?? ""

                if width == 0 || height == 0 {
                    _badArgumentError(result)
                    return
                }

                dispatchQueue.async {
                    var snapshotsData: [Any] = []
                    let data: [String: Any] = ["width": width, "height": height, "href": screen]

                    let snapshotData: [String: Any] = ["type": 4, "data": data, "timestamp": timestamp]
                    snapshotsData.append(snapshotData)

                    PaylisherSDK.shared.capture("$snapshot", properties: ["$snapshot_source": "mobile", "$snapshot_data": snapshotsData])
                }

                result(nil)
            } else {
                _badArgumentError(result)
            }
        #else
            result(nil)
        #endif
    }

    private func sendFullSnapshot(_ call: FlutterMethodCall,
                                  result: @escaping FlutterResult)
    {
        #if os(iOS)
            let date = Date()
            let timestamp = dateToMillis(date)
            if let args = call.arguments as? [String: Any] {
                let id = args["id"] as? Int ?? 1
                let x = args["x"] as? Int ?? 0
                let y = args["y"] as? Int ?? 0

                guard let imageBytes = args["imageBytes"] as? FlutterStandardTypedData else {
                    _badArgumentError(result)
                    return
                }

                dispatchQueue.async {
                    guard let image = UIImage(data: imageBytes.data) else {
                        return
                    }

                    guard let base64 = imageToBase64(image) else {
                        return
                    }

                    var snapshotsData: [Any] = []
                    var wireframes: [Any] = []

                    let wireframe: [String: Any] = [
                        "id": id,
                        "x": x,
                        "y": y,
                        "width": Int(image.size.width),
                        "height": Int(image.size.height),
                        "type": "screenshot",
                        "base64": base64,
                        "style": [:],
                    ]

                    wireframes.append(wireframe)
                    let initialOffset = ["top": 0, "left": 0]
                    let data: [String: Any] = ["initialOffset": initialOffset, "wireframes": wireframes]
                    let snapshotData: [String: Any] = ["type": 2, "data": data, "timestamp": timestamp]
                    snapshotsData.append(snapshotData)

                    PaylisherSDK.shared.capture("$snapshot", properties: ["$snapshot_source": "mobile", "$snapshot_data": snapshotsData])
                }

                result(nil)
            } else {
                _badArgumentError(result)
            }
        #else
            result(nil)
        #endif
    }

    private func isSessionReplayActive(result: @escaping FlutterResult) {
        #if os(iOS)
            result(PaylisherSDK.shared.isSessionReplayActive())
        #else
            result(false)
        #endif
    }

    private func openUrl(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let url = call.arguments as? String,
           let urlObject = URL(string: url)
        {
            #if os(iOS)
                if UIApplication.shared.canOpenURL(urlObject) {
                    UIApplication.shared.open(urlObject)
                }
            #else
                NSWorkspace.shared.open(urlObject)
            #endif
            result(nil)
        } else {
            result(FlutterError(code: "InvalidArguments",
                                message: "Invalid URL",
                                details: "The URL provided is invalid"))
        }
    }

    private func setup(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any] {
            PaylisherFlutterPlugin.setupPaylisher(args)
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func getFeatureFlag(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let featureFlagKey = args["key"] as? String
        {
            let value = PaylisherSDK.shared.getFeatureFlag(featureFlagKey)
            result(value)
        } else {
            _badArgumentError(result)
        }
    }

    private func isFeatureEnabled(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let featureFlagKey = args["key"] as? String
        {
            let value = PaylisherSDK.shared.isFeatureEnabled(featureFlagKey)
            result(value)
        } else {
            _badArgumentError(result)
        }
    }

    private func getFeatureFlagPayload(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let featureFlagKey = args["key"] as? String
        {
            let value = PaylisherSDK.shared.getFeatureFlagPayload(featureFlagKey)
            result(value)
        } else {
            _badArgumentError(result)
        }
    }

    private func identify(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let userId = args["userId"] as? String
        {
            let userProperties = args["userProperties"] as? [String: Any]
            let userPropertiesSetOnce = args["userPropertiesSetOnce"] as? [String: Any]

            PaylisherSDK.shared.identify(
                userId,
                userProperties: userProperties,
                userPropertiesSetOnce: userPropertiesSetOnce
            )
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func capture(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let eventName = args["eventName"] as? String
        {
            let properties = args["properties"] as? [String: Any]
            PaylisherSDK.shared.capture(
                eventName,
                properties: properties
            )
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func screen(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let screenName = args["screenName"] as? String
        {
            let properties = args["properties"] as? [String: Any]
            PaylisherSDK.shared.screen(
                screenName,
                properties: properties
            )
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func alias(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let alias = args["alias"] as? String
        {
            PaylisherSDK.shared.alias(alias)
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func distinctId(_ result: @escaping FlutterResult) {
        let val = PaylisherSDK.shared.getDistinctId()
        result(val)
    }

    private func reset(_ result: @escaping FlutterResult) {
        PaylisherSDK.shared.reset()
        result(nil)
    }

    private func enable(_ result: @escaping FlutterResult) {
        PaylisherSDK.shared.optIn()
        result(nil)
    }

    private func disable(_ result: @escaping FlutterResult) {
        PaylisherSDK.shared.optOut()
        result(nil)
    }

    private func isOptOut(_ result: @escaping FlutterResult) {
        let isOptedOut = PaylisherSDK.shared.isOptOut()
        result(isOptedOut)
    }

    private func debug(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let debug = args["debug"] as? Bool
        {
            PaylisherSDK.shared.debug(debug)
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func reloadFeatureFlags(_ result: @escaping FlutterResult
    ) {
        PaylisherSDK.shared.reloadFeatureFlags()
        result(nil)
    }

    private func group(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let groupType = args["groupType"] as? String,
           let groupKey = args["groupKey"] as? String
        {
            let groupProperties = args["groupProperties"] as? [String: Any]
            PaylisherSDK.shared.group(type: groupType, key: groupKey, groupProperties: groupProperties)
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func register(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let key = args["key"] as? String,
           let value = args["value"]
        {
            PaylisherSDK.shared.register([key: value])
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func unregister(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? [String: Any],
           let key = args["key"] as? String
        {
            PaylisherSDK.shared.unregister(key)
            result(nil)
        } else {
            _badArgumentError(result)
        }
    }

    private func flush(_ result: @escaping FlutterResult) {
        PaylisherSDK.shared.flush()
        result(nil)
    }

    private func captureException(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for captureException", details: nil))
            return
        }

        let properties = arguments["properties"] as? [String: Any]

        PaylisherSDK.shared.capture("$exception", properties: properties)
        result(nil)
    }

    private func close(_ result: @escaping FlutterResult) {
        PaylisherSDK.shared.close()
        result(nil)
    }

    private func getSessionId(result: @escaping FlutterResult) {
        result(PaylisherSDK.shared.getSessionId())
    }

    // Return bad Arguments error
    private func _badArgumentError(_ result: @escaping FlutterResult) {
        result(FlutterError(
            code: "PaylisherFlutterException", message: "Missing arguments!", details: nil
        ))
    }

    private func invokeFlutterMethod(_ method: String, arguments: Any? = nil) {
        DispatchQueue.main.async { [weak self] in
            self?.channel?.invokeMethod(method, arguments: arguments)
        }
    }

    private func requestNotificationPermission(_ result: @escaping FlutterResult) {
        #if os(iOS)
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) { granted, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "PERMISSION_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil) // success implies request completed, granted status not returned in existing interface? 
                    // Wait, existing interface Future<void> requestNotificationPermission().
                    // So returning nil is correct for success/completion.
                }
            }
        }
        #else
        result(nil)
        #endif
    }
}

// MARK: - PaylisherDeepLinkHandler & Application Life Cycle

extension PaylisherFlutterPlugin: PaylisherDeepLinkHandler {
    
    public func application(_ application: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
        return PaylisherSDK.shared.handleDeepLink(url)
    }
    
    public func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]) -> Void) -> Bool {
        return PaylisherSDK.shared.handleUserActivity(userActivity)
    }
    
    public func paylisherDidReceiveDeepLink(_ deepLink: PaylisherDeepLink, requiresAuth: Bool) {
        let event: [String: Any] = [
            "event": "deepLinkReceived",
            "data": [
                "url": deepLink.url.absoluteString,
                "scheme": deepLink.scheme ?? "",
                "destination": deepLink.destination,
                "parameters": deepLink.parameters ?? [:],
                "campaignId": deepLink.campaignId ?? ""
            ]
        ]
        
        PaylisherFlutterNotificationManager.shared.sendEvent(event)
    }
    
    public func paylisherDeepLinkRequiresAuth(_ deepLink: PaylisherDeepLink, completion: @escaping (Bool) -> Void) {
        // Auto-complete for now as no Dart callback mechanism is defined for auth flow
        completion(true)
    }
    
    public func paylisherDeepLinkDidFail(_ url: URL, error: Error?) {
        print("[PaylisherFlutter] Deep link failed: \(url), error: \(String(describing: error))")
    }
}

// MARK: - Helper Functions

func dateToMillis(_ date: Date) -> Int64 {
    return Int64(date.timeIntervalSince1970 * 1000)
}

func imageToBase64(_ image: UIImage) -> String? {
    return image.jpegData(compressionQuality: 0.8)?.base64EncodedString()
}